src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,690 @@
+        /// and replace its m_WriteList arraylist field.
+        /// </summary>
+        private sealed class ConnectionArrayList : ArrayListWrapper
+        {
+            public ConnectionArrayList(ArrayList list) : base(list)
+            {
+            }
+
+            public override int Add(object value)
+            {
+                if (s_connectionType.IsInstanceOfType(value))
+                {
+                    // Replace the HttpWebRequest arraylist inside this Connection object,
+                    // which allows us to intercept each new HttpWebRequest object added under
+                    // this Connection.
+                    ArrayList originalArrayList = s_writeListField.GetValue(value) as ArrayList;
+                    HttpWebRequestArrayList newArrayList = new HttpWebRequestArrayList(originalArrayList ?? new ArrayList());
+
+                    s_writeListField.SetValue(value, newArrayList);
+                }
+
+                return base.Add(value);
+            }
+        }
+
+        /// <summary>
+        /// Helper class used for Connection.m_WriteList. The goal here is to
+        /// intercept all new HttpWebRequest objects being added to Connection.m_WriteList
+        /// and notify the listener about the HttpWebRequest that's about to send a request.
+        /// It also intercepts all HttpWebRequest objects that are about to get removed from
+        /// Connection.m_WriteList as they have completed the request.
+        /// </summary>
+        private sealed class HttpWebRequestArrayList : ArrayListWrapper
+        {
+            public HttpWebRequestArrayList(ArrayList list) : base(list)
+            {
+            }
+
+            public override int Add(object value)
+            {
+                HttpWebRequest request = value as HttpWebRequest;
+                if (request != null)
+                {
+                    s_instance.RaiseRequestEvent(request);
+                }
+
+                return base.Add(value);
+            }
+
+            public override void RemoveAt(int index)
+            {
+                HttpWebRequest request = base[index] as HttpWebRequest;
+                if (request != null)
+                {
+                    HttpWebResponse response = s_httpResponseAccessor(request);
+                    if (response != null)
+                    {
+                        s_instance.RaiseResponseEvent(request, response);
+                    }
+                }
+
+                base.RemoveAt(index);
+            }
+        }
+
+        #endregion
+
+        #region private methods
+
+        /// <summary>
+        /// Private constructor. This class implements a singleton pattern and only this class is allowed to create an instance.
+        /// </summary>
+        private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
+        {
+        }
+
+        private void RaiseRequestEvent(HttpWebRequest request)
```

Comment:
I can speak for this. :)
Since this is not a long term solution, we are taking the approach of getting just the hooks in place here and leaving the activity logic in the caller. There are two reasons:
1. For a single HttpWebRequest, it actually triggers multiple send and receive events. An example is HTTP redirection. DiagnosticSource doesn't have full information to deduplicate these events. Application Insights however, has enough states to handle duplicates and log only one AI event.
2. Application Insights has a path handling cases where user uses profiler, so it handles all the activity stuff already. It's far easier to share that code path with this new path.

In the long term, we can look at adding activity logic here. However, I kind of like to first see how well this change goes.